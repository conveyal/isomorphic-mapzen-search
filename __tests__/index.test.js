/* globals describe, expect, it */

const nock = require('nock')

const geocoder = require('../index')

const mockReverseResult = require('./mock-reverse-result.json')
const mockSearchResult = require('./mock-search-result.json')
const mockKey = 'test-key'

describe('autocomplete', () => {
  it('should successfully autocomplete', async () => {
    nock('https://search.mapzen.com/')
      .get(/v1\/autocomplete/)
      .reply(200, mockSearchResult)

    const result = await geocoder.autocomplete({
      apiKey: mockKey,
      text: '123 a'
    })
    expect(result.features[0].geometry.coordinates[0]).toEqual(-77.023104)
  })

  it('should successfully autocomplete with custom headers', async () => {
    nock('https://search.mapzen.com/', {
      // Check for custom headers passed in below autocomplete request.
      reqheaders: {
        'x-api-key': (headerValue) => headerValue.includes('123')
      }
    })
      .get(/v1\/autocomplete/)
      .reply(200, mockSearchResult)

    const result = await geocoder.autocomplete({
      options: { headers: { 'x-api-key': 'abc123' } },
      text: '123+Main+St'
    })
    expect(result.features[0].geometry.coordinates[0]).toEqual(-77.023104)
  })

  it('should successfully autocomplete and include query in response', async () => {
    nock('https://search.mapzen.com/')
      .get(/v1\/autocomplete/)
      .reply(200, mockSearchResult)

    const result = await geocoder.autocomplete({
      apiKey: mockKey,
      includeQueryInResponse: true,
      text: '123 a'
    })
    expect(result).toMatchSnapshot()
  })
})

describe('search', () => {
  const searchQuery = '123 abc st'

  it('should successfully geocode', async () => {
    nock('https://search.mapzen.com/')
      .get(/v1\/search/)
      .reply(200, mockSearchResult)

    const result = await geocoder.search({ apiKey: mockKey, text: searchQuery })
    expect(result.features[0].geometry.coordinates[0]).toEqual(-77.023104)
  })

  it('should successfully geocode with a custom url', async () => {
    nock('https://custom.search.com/')
      .get(/v1\/search/)
      .reply(200, mockSearchResult)

    const result = await geocoder.search({
      apiKey: mockKey,
      text: searchQuery,
      url: 'https://custom.search.com/v1/search'
    })
    expect(result.features[0].geometry.coordinates[0]).toEqual(-77.023104)
  })

  it('should successfully geocode and format', async () => {
    nock('https://search.mapzen.com/')
      .get(/v1\/search/)
      .reply(200, mockSearchResult)

    const result = await geocoder.search({
      apiKey: mockKey,
      format: true,
      text: searchQuery
    })
    expect(result).toMatchSnapshot()
    expect(result[0].address).toEqual('Takoma, Takoma Park, MD, USA')
  })
})

describe('reverse', () => {
  const reverseQuery = { lat: 38.976745, lng: -77.023104 }

  it('should successfully reverse geocode', async () => {
    nock('https://search.mapzen.com/')
      .get(/v1\/reverse/)
      .reply(200, mockReverseResult)

    const result = await geocoder.reverse({
      apiKey: mockKey,
      point: reverseQuery
    })
    expect(result.features[0].geometry.coordinates[0]).toEqual(-77.023104)
  })

  it('should successfully reverse geocode and format', async () => {
    nock('https://search.mapzen.com/')
      .get(/v1\/reverse/)
      .reply(200, mockReverseResult)

    const result = await geocoder.reverse({
      apiKey: mockKey,
      format: true,
      point: reverseQuery
    })
    expect(result).toMatchSnapshot()
    expect(result[0].address).toEqual('Takoma, Takoma Park, MD, USA')
  })
})
