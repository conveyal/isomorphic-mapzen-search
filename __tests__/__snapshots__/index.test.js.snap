// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autocomplete should successfully autocomplete and include query in response 1`] = `
Object {
  "bbox": Array [
    -77.017897,
    38.904,
    -76.978642,
    39.001084,
  ],
  "features": Array [
    Object {
      "bbox": Array [
        -77.017242,
        38.973896,
        -77.012062,
        38.980594,
      ],
      "geometry": Object {
        "coordinates": Array [
          -77.023104,
          38.976745,
        ],
        "type": "Point",
      },
      "properties": Object {
        "accuracy": "centroid",
        "confidence": 0.965,
        "country": "United States",
        "country_a": "USA",
        "country_gid": "whosonfirst:country:85633793",
        "county": "Montgomery County",
        "county_gid": "whosonfirst:county:102082719",
        "gid": "whosonfirst:neighbourhood:85851759",
        "id": "85851759",
        "label": "Takoma, Takoma Park, MD, USA",
        "layer": "neighbourhood",
        "locality": "Takoma Park",
        "locality_gid": "whosonfirst:locality:85949501",
        "name": "Takoma",
        "neighbourhood": "Takoma",
        "neighbourhood_gid": "whosonfirst:neighbourhood:85851759",
        "region": "Maryland",
        "region_a": "MD",
        "region_gid": "whosonfirst:region:85688501",
        "source": "whosonfirst",
        "source_id": "85851759",
      },
      "type": "Feature",
    },
  ],
  "geocoding": Object {
    "attribution": "https://search.mapzen.com/v1/attribution",
    "engine": Object {
      "author": "Mapzen",
      "name": "Pelias",
      "version": "1.0",
    },
    "query": Object {
      "boundary.circle.lat": 38.8886,
      "boundary.circle.lon": -77.043,
      "boundary.circle.radius": 25,
      "private": false,
      "querySize": 20,
      "size": 10,
      "sources": Array [
        "geonames",
        "openaddresses",
        "openstreetmap",
        "whosonfirst",
      ],
      "text": "takoma",
    },
    "timestamp": 1479272483487,
    "version": "0.2",
  },
  "isomorphicMapzenSearchQuery": Object {
    "api_key": "test-key",
    "sources": "gn,oa,osm,wof",
    "text": "123 a",
  },
  "type": "FeatureCollection",
}
`;

exports[`reverse should successfully reverse geocode and format 1`] = `
Array [
  Object {
    "accuracy": "centroid",
    "address": "Takoma, Takoma Park, MD, USA",
    "confidence": 1,
    "country": "United States",
    "country_a": "USA",
    "country_gid": "whosonfirst:country:85633793",
    "county": "Montgomery County",
    "county_gid": "whosonfirst:county:102082719",
    "distance": 0,
    "gid": "whosonfirst:neighbourhood:85851759",
    "id": "85851759",
    "label": "Takoma, Takoma Park, MD, USA",
    "latlng": Object {
      "lat": 38.976745,
      "lon": -77.023104,
    },
    "layer": "neighbourhood",
    "locality": "Takoma Park",
    "locality_gid": "whosonfirst:locality:85949501",
    "name": "Takoma",
    "neighbourhood": "Takoma",
    "neighbourhood_gid": "whosonfirst:neighbourhood:85851759",
    "region": "Maryland",
    "region_a": "MD",
    "region_gid": "whosonfirst:region:85688501",
    "source": "whosonfirst",
    "source_id": "85851759",
  },
]
`;

exports[`search should successfully geocode and format 1`] = `
Array [
  Object {
    "accuracy": "centroid",
    "address": "Takoma, Takoma Park, MD, USA",
    "confidence": 0.965,
    "country": "United States",
    "country_a": "USA",
    "country_gid": "whosonfirst:country:85633793",
    "county": "Montgomery County",
    "county_gid": "whosonfirst:county:102082719",
    "gid": "whosonfirst:neighbourhood:85851759",
    "id": "85851759",
    "label": "Takoma, Takoma Park, MD, USA",
    "latlng": Object {
      "lat": 38.976745,
      "lon": -77.023104,
    },
    "layer": "neighbourhood",
    "locality": "Takoma Park",
    "locality_gid": "whosonfirst:locality:85949501",
    "name": "Takoma",
    "neighbourhood": "Takoma",
    "neighbourhood_gid": "whosonfirst:neighbourhood:85851759",
    "region": "Maryland",
    "region_a": "MD",
    "region_gid": "whosonfirst:region:85688501",
    "source": "whosonfirst",
    "source_id": "85851759",
  },
]
`;
